Complete all tasks.

JohnGudron opened the pull request:
Поменял код, подробно в комментарии к новому коммиту.
Новый пулл реквест делать не стал, на будущее спросил в чате, как лучше. Если надо будет - буду делать новые)

JohnGudron committed:
Complete all tasks.

kekulta commented on src/main/java/Race.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    public String findWinner() {
+        cars.sort(Comparator.comparing(car -> car.speed * 24));
```
Очень интересное решение для первого спринта

kekulta commented on src/main/java/Race.java:
```diff
@@ -0,0 +1,29 @@
+public class Race {
+
+    Scanner scanner = new Scanner(System.in);
+    ArrayList<Car> cars = new ArrayList<>();
```
Обычно из классов не торчат переменные, только методы.

JohnGudron replied:
А как лучше сделать - инициализировать сканер в main методе и передать его в конструктор Race класса или просто полностью вынести работу со Scanner в main метод? 
Или вопрос в private просто и добавлении сеттера для cars?

kekulta replied:
Сканер лучше иметь один на приложение. Можно передавать его внутрь класса, в этом нет ничего плохого.

kekulta replied:
Да, изначальный комментарий был про сеттеры/геттеры.

kekulta commented on src/main/java/Race.java:
```diff
@@ -0,0 +1,29 @@
+    Scanner scanner = new Scanner(System.in);
+    ArrayList<Car> cars = new ArrayList<>();
+
+    public Car takeInput(String carNum) {
+        System.out.printf("Введите через пробел имя и скорость %s автомобиля. Обращаем внимание, что скорость не может быть меньше 1 и больше 250.%n", carNum);
+        String[] input = scanner.nextLine().split(" ");
+        if (!checkSpeed(input[1])) {
+            System.out.println("Введена неверная скорость, повторите ввод");
+            takeInput(carNum);
+        }
+        return new Car(input[0], Integer.parseInt(input[1]));
+    }
```
Внутри этой же функции можно добавлять новую машину в список.

JohnGudron replied:
Логично)
А рекурсию так использовать норм или не стоит, чтобы на Stack over flow не нарваться?

kekulta replied:
В целом рекурсии лучше избегать, но конкретно в этом случае нормально, т.к. очень маловероятно, что стек уйдет глубоко.

JohnGudron committed:
Change Car class:
 - make fields private
 - add getter for fields
 - p.s. don't know about record classes, so left it as it is

Change Race class:
 - make fields private

Change Race.takeInput:
 - add check for appropriate input (not only speed)
 - fix previous recursive call, cause it leads to crashes - make while instead
 - new recursive call (added for new check) have "return", so stack collapsing as it should

Change Race.checkSpeed:
 - add first check via regex

kekulta commented on src/main/java/Race.java:
```diff
@@ -4,26 +4,38 @@
 
     public static boolean checkSpeed(String speed) {
-        return Integer.parseInt(speed) > 0 && Integer.parseInt(speed) <= 250;
+        if (!(speed.matches("\\d+"))) {
```
Я бы лучше использовал что-то вроде такого: https://stackoverflow.com/questions/8391979/does-java-have-a-int-tryparse-that-doesnt-throw-an-exception-for-bad-data

JohnGudron replied:
Да, хотелось что-то типа toIntOrNull(), но сходу не нашел, решил так. Еще лишние скобки зачем-то)

kekulta replied:
В котоине будет такое, в джаве нет, насколько я помню. В целом реджексы лучше аккуратно использовать, потому что довольно дорогая вещь.

JohnGudron replied:
Не задумывался про затратность реджексов - буду знать, спасибо. В целом, да, логичнее было бы трай кеч (раз строгих ограничений на имя нету) и завернуть бы его в экстеншн, для красоты)
